pkg/services/sqlstore/migrations/correlations_mig.go

```diff
@@ -59,4 +59,8 @@ func addCorrelationsMigrations(mg *Migrator) {
 		"description": "description",
 		"config":      "config",
 	})
+
+	mg.AddMigration("add provisioning column", NewAddColumnMigration(correlationsV1, &Column{
+		Name: "provisioned", Type: DB_Bool, Nullable: false, Default: "0",
```

Comment:
Not sure this would work if the table already includes rows?

Reply:
Yes, it will not correctly mark existing records. Existing correlations may be created manually or be provisioned. However, I don't know if there's a way to figure out which ones are actually provisioned or not (there's no indication if the source data source is actually provisioned or not).

The idea to approach it is treat all existing records as not-provisioned, meaning that **even if a correlation was provisioned, it will be duplicated upon restart, marked as provisioned and the user would need to manually remove non-provisioned one. The duplication may happen only once, with no more duplicates piling up because: during provisioning all previously provisioned correlations are removed** (I should've added it to the description and changelog notes).

We do not expect a lot of problems related to duplicated entries. It's not possible to provision correlations in Cloud so no Cloud customers are affected. It's still early preview / opt-in features so we expect very limited impact for on-prem customers. 

However, alternatively we can:
- introduce some de-dup logic (e.g. on load) and remove duplicates based on equality of each column, or 
- add API and manual way to trigger de-duplication, or
- add a background clean-up process.

Adding any of the above however, may mean we need to maintain this code for long time, and I wasn't sure it worth adding if it has very low impact.

Reply:
Not sure you understood exactly what I meant. Historically we haven't been able to add a new column with nullable false and a default value in case there's been existing data in the table. Not sure if this applies to mysql or postgres, but maybe not sqlite.

For example for uid column added to the datasource table:
https://github.com/grafana/grafana/blob/364ca3daa0eab69bcec9f2b70cb08ee8b69e5ab3/pkg/services/sqlstore/migrations/datasource_mig.go#L117-L129

Reply:
Oh I see. I wasn't aware there might be problem with it. I'll double-check, test it and update migration if needed.

Reply:
@marefr I've checked Postgres (15.4) and MySQL (8.0.32) and everything seemed to work fine.  I tested it with this scenario:

* create a clean database in main, 
* add some records in `correlation` table, 
* switch the this PR branch
* restart the service to run migration

In both cases the migration worked as expected and the default value was set up in all existing records correctly.

I also tried running the following SQL statements in Postgres and MySQL:

```sql
create table test (foo varchar(50));
insert into test(foo) values ('foo1');
insert into test(foo) values ('foo2');
alter table test add column bar varchar(50) not null default 'bar';
```

The last statement (representing the expected statement of the migration) passed and the default value was set up.

I also found this conversation: https://github.com/grafana/grafana/pull/23585#discussion_r408764230, but I struggle to figure out the case when the migration would not work as expected. 

Do you know what am I missing here? 🤔 Was this issue related to a specific db version? Or maybe was related to creating a column with unique values? I can also change `nullable=true` just in case as it should be prevented in the code anyway, but would be good to understand it better.

Reply:
Hmm maybe it was when using older mysql version then

Reply:
I think this should go through a Hosted Grafana Feature Readiness Review in case you haven't considered that to get some additional operator eyes on this.